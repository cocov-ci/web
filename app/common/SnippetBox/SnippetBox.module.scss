@use "sass:map";
@import "styles/variables.module";

.base {
  display: flex;
  position: relative;
  background: map.get($colorBox, "background");
  border: 1px solid map.get($colorBox, "border");
  border-radius: 8px;
  font-size: 17px;
  color: $colorBlack;

  .multilineInput {
    resize: none;
    min-height: 130px;
  }

  .multilineInput,
  .singleLineInput {
    margin: 0;
    border: 0;
    background: transparent;
    display: block;
    padding: 16px 50px 16px 15px;
    flex: 1;

    &:focus {
      outline: 0;
    }
  }

  .copyButton {
    position: absolute;
    top: 8px;
    right: 8px;
    overflow: hidden;

    .iconWrapper {
      position: absolute;
      display: flex;
      align-items: center;
      justify-content: center;
      height: 100%;
      width: 100%;
      transition: transform 0.1s ease-in;

      &.check {
        top: 100%;
      }

      &.moveUp {
        transform: translateY(-100%);
      }
    }
  }
}
