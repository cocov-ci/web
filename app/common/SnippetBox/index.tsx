'use client'

import classNames from 'classnames'
import React, { useEffect, useState } from 'react'

import AccessoryButton, {
  CheckIcon,
  CopyIcon,
} from 'app/common/AccessoryButton'
import { inconsolata } from 'utils/fonts'

import styles from './SnippetBox.module.scss'

type SnippetBoxProps = {
  className?: string
  source: string
  multiline?: boolean
}

const SnippetBox = ({ className, source, multiline }: SnippetBoxProps) => {
  const autoSelect = (
    ev: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>,
  ) => {
    ev.target.select()
  }

  const [showingCheck, setShowingCheck] = useState<boolean>(false)
  const [checkTimer, setCheckTimer] = useState<null | NodeJS.Timeout>(null)
  useEffect(() => {
    if (!checkTimer && showingCheck) {
      const timer = setTimeout(() => setShowingCheck(false), 2000)
      setCheckTimer(timer)
    } else if (checkTimer && !showingCheck) {
      clearTimeout(checkTimer)
      setCheckTimer(null)
    }

    return () => {
      if (checkTimer) {
        clearTimeout(checkTimer)
        setCheckTimer(null)
      }
    }
  }, [showingCheck])

  const doCopy = () => {
    if (!window.navigator) {
      return
    }

    window.navigator.clipboard
      .writeText(source)
      .then(() => setShowingCheck(true))
  }

  return (
    <div
      className={classNames(styles.base, className, {
        [styles.multiline]: multiline === true,
      })}
    >
      {multiline === true ? (
        <textarea
          className={classNames(styles.multilineInput, inconsolata.className)}
          onFocus={autoSelect}
          readOnly={true}
          value={source}
        />
      ) : (
        <input
          className={classNames(styles.singleLineInput, inconsolata.className)}
          onFocus={autoSelect}
          readOnly={true}
          type="text"
          value={source}
        />
      )}
      <AccessoryButton
        className={styles.copyButton}
        kind="squared"
        onClick={() => doCopy()}
        title="Copy"
      >
        <div
          className={classNames(styles.iconWrapper, {
            [styles.moveUp]: showingCheck,
          })}
        >
          <CopyIcon />
        </div>
        <div
          className={classNames(styles.iconWrapper, styles.check, {
            [styles.moveUp]: showingCheck,
          })}
        >
          <CheckIcon />
        </div>
      </AccessoryButton>
    </div>
  )
}

export default SnippetBox
