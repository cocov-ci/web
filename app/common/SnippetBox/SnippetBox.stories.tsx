import type { Meta, StoryObj } from '@storybook/react'

import SnippetBox from './index'

const meta: Meta<typeof SnippetBox> = {
  title: 'Common/SnippetBox',
  component: SnippetBox,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof SnippetBox>

export const SingleLine: Story = {
  args: {
    source: 'crt_d3e4ad8b6735a2fa04f546f68b51df73039463df16',
  },
}

export const MultiLine: Story = {
  args: {
    source: `<a href=“https://cocov.company.io/repos/jps/coverage”><img alt="" src="https://badges-cocov.company.io/api/coverage” /></a>`,
    multiline: true,
  },
}
