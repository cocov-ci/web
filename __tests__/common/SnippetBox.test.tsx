import { fireEvent, render, screen } from '@testing-library/react'

import SnippetBox from 'app/common/SnippetBox'

import '@testing-library/jest-dom'

jest.useFakeTimers()
jest.spyOn(global, 'setTimeout')

Object.assign(navigator, {
  clipboard: {
    writeText: jest.fn(() => Promise.resolve()),
  },
})

describe('common/SnippetBox', () => {
  it('renders SnippetBox', () => {
    render(
      <SnippetBox source="crt_d3e4ad8b6735a2fa04f546f68b51df73039463df16" />,
    )

    expect(screen.getByRole('textbox')).toHaveValue(
      'crt_d3e4ad8b6735a2fa04f546f68b51df73039463df16',
    )
  })

  it('triggers autoSelect event', () => {
    const handleEvent = jest.fn()

    render(
      <SnippetBox source="crt_d3e4ad8b6735a2fa04f546f68b51df73039463df16" />,
    )

    fireEvent.focus(screen.getByRole('textbox'), {
      target: { select: handleEvent },
    })

    expect(handleEvent).toHaveBeenCalledTimes(1)
  })

  it('renders SnippetBox snapshots', () => {
    const { container } = render(
      <SnippetBox source="crt_d3e4ad8b6735a2fa04f546f68b51df73039463df16" />,
    )

    expect(container).toMatchSnapshot()
  })
})
