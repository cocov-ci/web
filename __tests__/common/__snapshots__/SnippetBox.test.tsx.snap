// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/SnippetBox renders SnippetBox snapshots 1`] = `
<div>
  <div
    class="base"
  >
    <input
      class="singleLineInput className"
      readonly=""
      type="text"
      value="crt_d3e4ad8b6735a2fa04f546f68b51df73039463df16"
    />
    <button
      class="base copyButton squared"
      title="Copy"
    >
      <div
        class="iconWrapper"
      >
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="13"
          loading="lazy"
          src="/icons/copy.svg"
          srcset="/icons/copy.svg 1x, /icons/copy.svg 2x"
          style="color: transparent;"
          width="13"
        />
      </div>
      <div
        class="iconWrapper check"
      >
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="11"
          loading="lazy"
          src="/icons/check.svg"
          srcset="/icons/check.svg 1x, /icons/check.svg 2x"
          style="color: transparent;"
          width="16"
        />
      </div>
    </button>
  </div>
</div>
`;
